### DISCUSSION TITLE
Features List

### DISCUSSION THREAD

**Discussion by ckipp01** [scalameta/nvim-metals#279] [Announcement]
This is an ongoing demonstration of features that exist while using `nvim-metals`.

Feel free to subscribe to this to get notified when new features are added!

A couple things to keep in mind when looking at the examples.

1. I'm using some other plugins which might change the appearance of the defaults. For example, I use [`telescope.nvim`](https://github.com/nvim-telescope/telescope.nvim) along with the telescope extension `nvim-metals` provides to trigger all my metals commands. All commands are available this way. They are also all available with their `:command` equivalents. For example `:MetalsImportBuild` is the same as the Import Build command in the telescope extension.
2. I use [`nvim-bqf`](https://github.com/kevinhwang91/nvim-bqf) to give a nice preview of what's in my quickfix. So commands that will by default pipe stuff into my quickfix window will then be automatically previewed.

Again, this is just my workflow. The beauty of the built-in LSP of Neovim is that you can mix and match all sorts of stuff to get the exact experience you want. The above extensions work great for me, but maybe not for you. You'll have to experiment or just use the default, that works too.

### Features

- [Importing your build](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1751981)
- [Triggering specific compiles](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752011)
- [Goto definition](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752019)
- [Find references](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752027)
- [Find implementations](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752032)
- [Hovers (plus inferred types, implicit params, and other synthetics)](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752060)
- [Goto super method](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752098)
- [Rename](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752116)
- [Insert type annotation](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752121)
- [Organize imports](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752134)
- [Formatting](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752192)
- [Signature help](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752218)
- [Document symbols](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752237)
- [Workspace symbols](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752249)
- [Workspace diagnostics](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752266)
- [Switching your build server](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752276)
- [Worksheet support](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752305)
- [Analyzing stacktraces](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752360)
- [Find in dependency jars (useful for searching configs)](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1753317)
- [Generating a BSP config](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1753330)
- [Info about Metals, locations of files, etc](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1753707)
- [Creating a new Scala file](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1753724)
- [TVP Support](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1774629)
- [Show various formats of the file](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1774680)
- [Debugging support](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1774771)
- [Creating a new Scala project](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1774852)
- [Creating a new Java file](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1883057)
- [Show build target info](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-2223593)
- [Toggling the Metals Logs](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-2825966)
- [Run Scalafix Rules](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-2928579)

**Comment by ckipp01** [comment]
# Importing your build

Upon opening a new project for any supported build tool, you'll see a prompt to import your build. This should happen automatically in a new workspace and also when you change something in your build. Keep in mind that if you make multiple changes it a row, it may not prompt you every time because there is a refresh time so we don't constantly spam you with requests if you're making a lot of changes in succession.

<img width="1130" alt="Screenshot 2021-12-05 at 12 30 20" src="https://user-images.githubusercontent.com/13974112/144744705-463700db-d736-449d-becc-ec5984a68ff8.png">

For some reason if you either ignore this request by saying `Not now` or `Don't show again` and you want to manually trigger that, you can using the `Import Build` command. (`:MetalsImportBuild`).

![2021-12-05 12 36 01](https://user-images.githubusercontent.com/13974112/144744825-4fc2bc05-1156-455d-ada0-86dd2bb08082.gif)

**Reply by arbitrary-dev** [reply]
How to trigger this "Metals Commands" popup? I don't see [a mapping](https://github.com/scalameta/nvim-metals/discussions/39#discussion-82302) for it or it being listed among available `:Metals*` commands...

**Reply by antosha417** [reply]
> How to trigger this "Metals Commands" popup? I don't see [a mapping](https://github.com/scalameta/nvim-metals/discussions/39#discussion-82302) for it or it being listed among available `:Metals*` commands...

it is done via [telescope](https://github.com/nvim-telescope/telescope.nvim) extensions. So you need to have it installed. Then try this command
`:lua require"telescope".extensions.metals.commands()`

**Reply by ghost** [reply]
Can I import a build with a specific heap size?

**Comment by ckipp01** [comment]
# Triggering specific compiles

Apart from the default save, which triggers a compile for your current build target and it's dependencies, you may also find situations where you want to not only compile those, but also the inverse dependencies. You can use a cascade compile for this (`:MetalsCompileCascade`).

Since cascades can take a while, there is also a Cancel Compilation command (`:MetalsCompileCancel`).

For those times where things may get out of wack, there is also a Compile Clean command to both clean and then compile (`:MetalsCompileClean`).

<img width="1130" alt="Screenshot 2021-12-05 at 12 48 11" src="https://user-images.githubusercontent.com/13974112/144745240-eabc2376-4052-4624-8dd5-586ec63d699f.png">

**Comment by ckipp01** [comment]
## Goto Definition

In order to use this you'll need a mapping for `vim.lsp.buf.definition()`. This works for both internal and external dependencies. You can also jump to java dependencies.

![2021-12-05 12 53 08](https://user-images.githubusercontent.com/13974112/144745420-4623a647-116f-40fa-b141-fd490a02c532.gif)

**Reply by arbitrary-dev** [reply]
But how does one open a class in `nvim-metals`?

In `coc-metals` it was: `<space>s<class_name><Enter>`

**Reply by ckipp01** [reply]
So looking at the config it looks like that calls `CocList -I symbols`, so the equivalent in `nvim-metals` would actually be be https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1752249. It might look a bit different depending on if you're using a plugin like telescope or not.

**Reply by ghost** [reply]
Is it enough if I have a map to `builtin.lsp_definitions` with Telescope? i.e. `vim.keymap.set('n', '<leader>gd', '<cmd>Telescope lsp_definitions<cr>')
`

**Reply by ckipp01** [reply]
If you want to use telescope for this, then sure.

**Comment by ckipp01** [comment]
# Find references

In order to use this you'll need a mapping for `vim.lsp.buf.references`.

![2021-12-05 12 55 24](https://user-images.githubusercontent.com/13974112/144745537-c18dd6a1-e75c-4bb0-a9cd-a2021a005328.gif)

**Reply by V4G4X** [reply]
In my config, the lower "quickfix list" window opens, but without the preview pane that hovers above.  
What could be the possible cause?

**Reply by ckipp01** [reply]
It's actually a plugin that I use that is showing the preview. It's called [nvim-bqf](https://github.com/kevinhwang91/nvim-bqf).

**Reply by V4G4X** [reply]
Thanks I got it.

**Reply by z4f1r0v** [reply]
When I select a reference, the window that popped up doesn't close. Do you have some special configuration for telescope, @ckipp01?

**Reply by ckipp01** [reply]
Hey @z4f1r0v I don't use Telescope for this feature so I'm not really sure. You may have to minimize it and report a bug to them if it's not closing correctly since this feature is actually just a core LSP feature that telescope hooks into unrelated to nvim-metals.

**Reply by mvillafuertem** [reply]
@ckipp01 I try to reproduce this same example, using a case class, but I get ‘No references found’, any thoughts?

```lua
vim.keymap.set("n", "gr", vim.lsp.buf.references)
```

MetalsRunDoctor
```text
## Metals Info
  - Metals Server version: 1.3.0
  - Metals Java: 17.0.3 from Azul Systems, Inc. located at /nix/store/znr41d41j7qp60y61lqcish0vhzwldmf-zulu17.34.19-ca-jdk-17.0.3/zulu-17.jdk/Contents/Home

Below are listed the build targets for this workspace. One build target corresponds to one classpath. For example, normally one sbt project maps to two build targets: main and test.

## Workspaces

### Workspace: /Users/user1/project/my-service
 - Build definition is coming from sbt.
 - Build server currently being used is Bloop v1.5.17.

## Build Targets

### my-service
  - target type: Scala 2.13.10
  - goto functionality: metalsDecode:file:///Users/user1/project/my-service/my-service.metals-buildtarget
  - compilation: ✅ 
  - diagnostics: ✅ 
  - interactive: ✅ 
  - semanticdb: ⚠️ 
  - debugging: ✅ 
  - java: ✅ 

```

**Comment by ckipp01** [comment]
# Find implementations

In order to use this you'll need a mapping for `vim.lsp.buf.implementation()`.

![2021-12-05 12 58 27](https://user-images.githubusercontent.com/13974112/144745614-273907b3-8e64-41c2-8daf-3aa315b4f109.gif)

**Reply by bblfish** [reply]
The trick to find out what to put in the lua config file is to run the `:Telescope` command which opens a list of the commands and a search field. For the `find implementations` examples shown above,  searching for "implement" returns `lsp_implementations` which I could then get to work by adding the following command in my `lua/plugins.lua` file:
```lua
vim.keymap.set("n", "gi", require("telescope.builtin").lsp_implementations )
```
after of course removing the default `gi` binding.

**Reply by mvillafuertem** [reply]
@ckipp01 any idea why it is failing?

```lua

{ "n", "gi", vim.lsp.buf.implementation, "Go to implementation" },

```

```lua 
Error executing vim.schedule lua callback: ...-unwrapped-0.11.5/share/nvim/runtime/lua/vim/lsp/buf.lua:226: Cursor position outside
 buffer
stack traceback:
        [C]: in function 'nvim_win_set_cursor'
        ...-unwrapped-0.11.5/share/nvim/runtime/lua/vim/lsp/buf.lua:226: in function 'on_response'
        ...-unwrapped-0.11.5/share/nvim/runtime/lua/vim/lsp/buf.lua:245: in function 'handler'
        ...wrapped-0.11.5/share/nvim/runtime/lua/vim/lsp/client.lua:682: in function ''
        vim/_editor.lua: in function <vim/_editor.lua:0>
```

```sh
metals --version
Checking available versions on https://repo1.maven.org/maven2/org/scalameta/

metals 1.6.3

```

I have tried this as well 
https://www.reddit.com/r/neovim/comments/1iwxejo/resolved_cursor_position_outside_buffer_when/

Similar errors

https://github.com/nvim-telescope/telescope.nvim/issues/3163#issuecomment-3175990343

**Comment by ckipp01** [comment]
# Hovers (plus inferred types, implicit params, and other synthetics)

In order to make this work you'll need a mapping for `vim.lsp.buf.hover()`.

If you'd also like the extra information shown in the `With synthetics added`, you'll also need to enable this in your metals config since they are off by default. See examples of this under the `metals-settings` section of the [`nvim-metals` docs](https://github.com/scalameta/nvim-metals/blob/main/doc/metals.txt).

<img width="1130" alt="Screenshot 2021-12-05 at 13 11 14" src="https://user-images.githubusercontent.com/13974112/144746081-da02ba4f-71f7-445c-a7a8-70c9cf7eb8b6.png">

**Comment by ckipp01** [comment]
# Goto super method

In order to use this you'll need to have a mapping for `vim.lsp.codelens.run()` and then trigger it where you see the code lens on an overridden member.

![2021-12-05 13 20 20](https://user-images.githubusercontent.com/13974112/144746260-56ff6cd0-eda5-4a6b-98fa-9c598cba4e85.gif)

In more advanced situations there is al an option to see the entire hierarchy and choose where you jump to using the Super Method Hierarchy command (`:MetalsSuperMethodHierarchy`).

![2021-12-05 13 22 14](https://user-images.githubusercontent.com/13974112/144746361-a78b4112-2e65-4d82-878d-668fc2019677.gif)

**Comment by ckipp01** [comment]
# Rename

In order to use this you'll need a mapping for `vim.lsp.buf.rename()`.

![2021-12-05 13 26 15](https://user-images.githubusercontent.com/13974112/144746480-4d7d0c8b-81e2-47ee-8094-3aa46c7d2b6b.gif)

**Comment by ckipp01** [comment]
# Insert type annotation

In order to use this you'll need a mapping for `vim.lsp.buf.code_action()`.

![2021-12-05 13 28 08](https://user-images.githubusercontent.com/13974112/144746548-6bc7cd89-5b7b-434a-9606-edcf1f5e89b8.gif)

**Comment by ckipp01** [comment]
# Organize imports

There are multiple ways to organize your imports. The first is using the Organize Imports command (`:MetalsOrganizeImports`).
![2021-12-05 13 32 31](https://user-images.githubusercontent.com/13974112/144746668-b0cd60e2-c2d2-4dda-9e19-c0bc5086748e.gif)

The second method is to be on top of an unused warning on an import and triggering a code action via your `vim.lsp.buf.code_action()` mapping.

![2021-12-05 13 33 41](https://user-images.githubusercontent.com/13974112/144746710-e7a700e0-423b-4fe0-9574-25246548162c.gif)

**Comment by ckipp01** [comment]
# Formatting

In order to use this you'll need a mapping for `vim.lsp.buf.formatting()`.

![2021-12-05 13 45 37](https://user-images.githubusercontent.com/13974112/144747100-c3c364a3-42c8-4aa7-b001-3703ee3bb1f2.gif)

**Comment by ckipp01** [comment]
# Signature help

In order to use this you'll need a mapping set up for `vim.lsp.buf.signature_help()`.

![2021-12-05 13 59 28](https://user-images.githubusercontent.com/13974112/144747579-1d55dcc5-f584-4c61-a600-1a962abd2a02.gif)

**Comment by ckipp01** [comment]
# Document symbols

In order to use this you'll need a mapping set up for `vim.lsp.buf.document_symbol()`. My gif below is utilizing [`telescope.nvim`](https://github.com/nvim-telescope/telescope.nvim) for this.

![2021-12-05 14 05 22](https://user-images.githubusercontent.com/13974112/144747825-e166ccbd-6b0a-444f-92c3-edadcb5260e2.gif)

**Reply by V4G4X** [reply]
How does your open line that? I have [nvim-bqf](https://github.com/kevinhwang91/nvim-bqf) installed so it uses that.  
But just for this command I'd love to have this window.

**Reply by ckipp01** [reply]
For this specific command I'm not actually using nvim-bqf, but rather telescope.nvim.

**Reply by V4G4X** [reply]
Got it. 
So basically it was `:Telescope` and then enter "Document" until the comands says `lsp_document_symbols`. 
Press enter and you get the above.  

Just mapped :Telescope to a keymapping so I get all the telescope command without having to learn each specific one.

**Reply by ckipp01** [reply]
Keep in mind that all these have lua api's as well. For example in my own config for this I use:

```lua
vim.keymap.set("n", "gds", require("telescope.builtin").lsp_document_symbols)
```

**Comment by ckipp01** [comment]
# Workspace symbols

In order to use this you'll need a mapping set up for `vim.lsp.buf.workspace_symbol()`. My gif below is utilizing [`telescope.nvim`](https://github.com/nvim-telescope/telescope.nvim) for this.


![2021-12-10 12 15 35](https://user-images.githubusercontent.com/13974112/145565741-4c80bf1f-5d3e-48c6-8d57-08387f7bbf61.gif)

A couple extra things to note for using this:

1. Metals has some cool extra things here. If you preface your search with `;` for example you can also search your external deps. You can also do shortened querys like `j.t.i` to search for `java.time.Instant`.
2. Depending on what you're using to display these results, some send in an empty query string to start off the process. Since this can potentially be a _huge_ amount of symbols, metals won't respond to an empty query search. So for example with telescope, I need to use `builtin.lsp_dynamic_workspace_symbols` not the normal `builtin.lsp_workspace_symbols`.

**Reply by bblfish** [reply]
I got the above to work when typing `gwz` by adding to my `lua/plugins.lua` file (after installing telescope and getting it to return a good health check with `:checkhealth telescope`
```lua
vim.keymap.set("n", "gwz",  require("telescope.builtin").lsp_dynamic_workspace_symbols )
```
just below the command shown in the default setup
```lua
map("n", "gws",  vim.lsp.buf.workspace_symbol )
```

**Comment by ckipp01** [comment]
# Workspace diagnostics

There are a lot of way to view your diagnostics as the `vim.diagnostic` api is pretty great. By default many use `vim.diagnostic.setloclist()`, but this only show you the diagnostics in the current buffer. I personally prefer to see all my workspace diagnostics with the ability to only see errors or warnings. You can do this using the following functions in your mappings:

```lua
vim.diagnostic.setqflist() // all diagnostics
vim.diagnostic.setqflist({severity = "E"}) // all errors
vim.diagnostic.setqflist({severity = "W"}) // all warnings
```

![2021-12-10 14 14 30](https://user-images.githubusercontent.com/13974112/145582677-e2db8324-7f48-4cde-8a70-bf6d6bd2d990.gif)

**Comment by ckipp01** [comment]
# Switching your build server

You can switch your build server by using the Switch Build Server command (`:MetalsSwitchBsp`).

![2021-12-05 14 22 17](https://user-images.githubusercontent.com/13974112/144748346-3701c280-9e90-406a-bab5-bccdfbcc3c88.gif)

**Comment by ckipp01** [comment]
# Worksheet support

By default opening any `*.worksheet.sc` should be evaluated by Metals as a worksheet. The evaluations will display on the right had side as output.

<img width="1130" alt="Screenshot 2021-12-05 at 14 26 31" src="https://user-images.githubusercontent.com/13974112/144748471-e239297d-d559-4f0f-a6cf-91672dd874c5.png">

## Long output or printed lines

If you have very long output or you maybe print lines in a function, you see this indicated by `…`. You can always see the full hover output by having a mapping set for `require("metals").hover_worksheet()`. Triggering this anywhere on the line will show you it.

![2021-12-05 14 29 08](https://user-images.githubusercontent.com/13974112/144748551-6828a274-b562-4a37-be91-1cca993b045b.gif)

## Copying all the worksheet output

Sometimes you want to be able to share some code with the output. You can do this by using the Copy Worksheet Output command (`:MetalsCopyWorksheetOutput`).

![2021-12-05 14 30 57](https://user-images.githubusercontent.com/13974112/144748605-d56b4149-b6c3-49f3-9a5a-6ef7c1889798.gif)

**Comment by ckipp01** [comment]
# Analyzing stacktraces

If you have a stack trace that you'd like to analyze it, just copy it and then trigger the Analyze Stacktrace command (`:MetalsAnalyzeStacktrace`). This will bring you to a `stacktrace.scala` file in your `.metals` dir that has code lenses that you can then trigger to jump to the parts of the trace. To utilize the code lens, you'll need to make sure you have a mapping set up for `vim.lsp.codelens.run()`.


![2021-12-05 14 51 40](https://user-images.githubusercontent.com/13974112/144749323-2e067150-ec57-44c9-b544-8371dbe4f471.gif)

**Comment by ckipp01** [comment]
# Find in dependency jars (useful for searching configs)

If you'd like to search through files in your dependency jars to for example find values in a config or another type of file, you can do this using the Find in Dependency Jars command (`:MetalsFindInDependencyJars`).

![2021-12-05 20 12 41](https://user-images.githubusercontent.com/13974112/144760291-72b213e1-428b-44ff-a5c5-740082a26494.gif)

**Comment by ckipp01** [comment]
# Generating a BSP config

If you're using a build tool that can also be a build server, such as sbt or mill, Metals can also generate the `.bsp/<entry>.json` file for that build tool. This is no longer necessary to do before switching via Switch Build Server, but if you'd like to ensure your BSP entry for this build tool is fresh you can use the Generate BSP Config command (`:MetalsGenerateBspConfig`).

<img width="1162" alt="Screenshot 2021-12-05 at 20 18 07" src="https://user-images.githubusercontent.com/13974112/144760393-e6a0d880-347b-466a-8342-2bfeed8fc669.png">

**Comment by ckipp01** [comment]
# Info about Metals, locations of files, etc

To see all this information you can use the Info command (`:MetalsInfo`).

![2021-12-05 22 43 42](https://user-images.githubusercontent.com/13974112/144765064-8dfe2192-4b62-410d-a82c-3a601b2dd456.gif)

You can also see information about all of your build targets and info about possibly misconfiguration in the server via the Metals Doctor command (`:MetalsRunDoctor`).

![2021-12-05 22 46 34](https://user-images.githubusercontent.com/13974112/144765151-c5800f81-8bd5-42a3-83e1-82af36dfa301.gif)

**Reply by amirphl** [reply]
It has changed to `:MetalsRunDoctor`.

**Comment by ckipp01** [comment]
# Creating a new Scala file

Metals can create new files which will include a skeleton structure for the type of file and also the package name. There are multiple ways to trigger this. The first is with the New Scala File command (`:MetalsNewScalaFile`).

![2021-12-05 22 53 03](https://user-images.githubusercontent.com/13974112/144765390-edb4824a-9d6e-4b51-9d37-9e2c20076f3b.gif)

The second way to trigger this is if you create a type that doesn't yet exist that you intend to make a file for, it will offer this to you when you trigger a code action via your `vim.lsp.buf.code_action()` mapping.

![2021-12-05 22 55 28](https://user-images.githubusercontent.com/13974112/144765445-f1f41f5f-dc36-44cd-91ea-dde570810739.gif)

**Comment by ckipp01** [comment]
# TVP Support

Metals has an LSP extension called the [Tree View Protocol](https://scalameta.org/metals/docs/integrations/tree-view-protocol) which can be used to provide various panels of information. `nvim-metals` implements  the `metalsPackages` portion of this. This provides a few different nice features. In order to use this you'll want mappings for `require("metals.tvp").toggle_tree_view()` and `require("metals.tvp").reveal_in_tree()`.

**NOTE**: There are some `tvp` config options that can be seen in `:h metals-tvp-module`, but the default ones you'll probably use the most is `<CR>` to expand a node and `r` to trigger the node command (jumping from tree into that place in the code).

### Jumping into the tree structure from your code and vice versa

![2021-12-08 19 04 53](https://user-images.githubusercontent.com/13974112/145260765-1f98acb9-83ac-4b3f-b6e4-36d520d4cd1f.gif)

NOTE: This also works for your external dependencies.

![2021-12-08 19 10 02](https://user-images.githubusercontent.com/13974112/145261023-40ac01a9-ad09-46c9-8a51-7c81307968a6.gif)

**Reply by arbitrary-dev** [reply]
Somehow for me it doesn't work. When I press Enter on any item in TVP – nothing happens.

**Reply by ckipp01** [reply]
> When I press Enter on any item in TVP – nothing happens.

Could you explain what you mean? Are you referring to the node command? How are you triggering it?

**Reply by arbitrary-dev** [reply]
Ah, I see now. It's not `Enter` to jump from node to code, but `r`...
Probably because `Enter` expands/collapses nodes that are also jumpable...
But I'd suggest adding `Enter` mapping for jumping to code from leafes as well, just for better UX.

**Reply by ckipp01** [reply]
> But I'd suggest adding Enter mapping for jumping to code from leafes as well, just for better UX.

But then you'd end up having to use `r` to make the jump for expandable nodes and `Enter` for those that don't expand. That sounds even more confusing.

**Reply by arbitrary-dev** [reply]
No, what I mean is to ADD `Enter` to non-expandable nodes.

So that it could be:
- For expandable nodes: `Enter` to expand, `r` to jump to code
- For non-expadable: `Enter` & `r` to jump to code

**Reply by ckipp01** [reply]
That's extra logic that I don't really feel adds much. I have the `node_command_mapping` in the docs, so if that's not clear I'd gladly accept a pr making it more clear, but I don't think I'll add the extra logic you mention.

**Reply by arbitrary-dev** [reply]
:)
```diff
 ## TVP Support
 
 Metals has an LSP extension called the [Tree View Protocol](https://scalameta.org/metals/docs/integrations/tree-view-protocol) which can be used to provide various panels of information. `nvim-metals` implements the `metalsPackages` portion of this. This provides a few different nice features. In order to use this you'll want mappings for `require("metals.tvp").toggle_tree_view()` and `require("metals.tvp").reveal_in_tree()`.
+
+Use `r` to jump from tree view to the code.
```

**Reply by antosha417** [reply]
> That's extra logic that I don't really feel adds much. I have the `node_command_mapping` in the docs, so if that's not clear I'd gladly accept a pr making it more clear, but I don't think I'll add the extra logic you mention.

I've used a couple of `"tree"` plugings in vim. And I think all of them open the file if you hit expand shortcut on something that is can't be expanded. For me this feature would add consistency among plugins. And it seems pretty straight forward to implement.

**Reply by arbitrary-dev** [reply]
>And it seems pretty straight forward to implement.

Contribute 😃

**Reply by antosha417** [reply]
you can try it out in a branch from this pr https://github.com/scalameta/nvim-metals/pull/440

**Comment by ckipp01** [comment]
# Show various formats of the file

Metals allows you to see your current file in various formats. The formats and their commands that are available are:

- Show Cfr (`:MetalsShowCfr`)
- Show Javap (`:MetalsShowJavap`)
- Show Javap Verbose (`:MetalsShowJavapVerbose`)
- Show Semanticdb Compact (`:MetalsShowSemanticdbCompact`)
- Show Semanticdb Detailed (`:MetalsShowSemanticdbDetailed`)
- Show Semanticdb Proto (`:MetalsShowSemanticdbProto`)
- Show Tasty (`:MetalsShowTasty`) NOTE: Only for Scala 3

![2021-12-08 19 14 38](https://user-images.githubusercontent.com/13974112/145262529-bef886b4-4738-42d9-80e1-91cab15193b3.gif)

**Comment by ckipp01** [comment]
# Debugging support

Thanks to the great work of [mfussenegger/nvim-dap](https://github.com/mfussenegger/nvim-dap) there is also support for running, testing, and debugging your code right in `nvim-metals`. For more information on this, please do check out the `nvim-dap` docs and also the section of docs for `nvim-metals` about the DAP integration. There are lots of extra plugins available to make the experience a bit fancier, but to give you a very general idea only using `nvim-dap`:

### Running your code

![2021-12-08 19 26 46](https://user-images.githubusercontent.com/13974112/145264981-86161cac-956a-4fcd-ba8b-93d81d37442f.gif)

### Setting a breakpoint and looking at the current scope

![2021-12-08 19 32 17](https://user-images.githubusercontent.com/13974112/145265060-d0de9cb4-3207-4469-89ad-b7307619beba.gif)

### Expression evaluation

![2021-12-08 19 33 49](https://user-images.githubusercontent.com/13974112/145265111-facd3b54-84a3-44ba-b045-cd7bbbfabaf0.gif)

### Running your tests

![2021-12-08 19 35 42](https://user-images.githubusercontent.com/13974112/145265178-7fdb3d77-c6cf-4380-b503-93c582b7869a.gif)

### Using code lenses

You can also directly trigger these via the `run`, `debug`, `test`, or `testDebug` code lenses.

![2021-12-14 18 04 14](https://user-images.githubusercontent.com/13974112/146046622-2ab85be3-ce9f-4a14-b136-77c977a509d6.gif)

**Reply by arbitrary-dev** [reply]
How do I get test output?

When I run `test` code lens nothing happens.

And I see these in logs:
```
2022.10.19 21:58:05 INFO  tracing is disabled for protocol dap-server, to enable trac
ing of incoming and outgoing JSON messages create an empty file at /home/user/project/.metals/dap-server.trace.json or /home/user/.cache/metals/da
p-server.trace.json
2022.10.19 21:58:05 INFO  tracing is disabled for protocol dap-client, to enable trac
ing of incoming and outgoing JSON messages create an empty file at /home/user/project/.metals/dap-client.trace.json or /home/user/.cache/metals/da
p-client.trace.json
2022.10.19 21:58:05 INFO  Starting debug proxy for [com.app.AppSpec]
2022.10.19 21:58:09 INFO  Trying to attach to remote debuggee VM localhost:60837 .
2022.10.19 21:58:09 INFO  Attaching to debuggee VM succeeded.
2022.10.19 21:58:15 INFO  Canceling debug proxy for [com.app.AppSpec]
2022.10.19 21:58:15 INFO  Closing debug server tcp://0.0.0.0:38255
```

**Reply by ckipp01** [reply]
The output will be in the DAP REPL. You can toggle it with `require("dap").repl.toggle()`.

**Reply by gogamid** [reply]
@ckipp01  Is there support for conditional breakpoint?

**Reply by linux-root** [reply]
Posting the code suggested by @ckipp01 here in case anyone needs it. This triggers the DAP REPL (Run Console) to open automatically after running code from nvim-metal, mimicking IntelliJ's behavior :
```lua
    dap.listeners.after["event_terminated"]["nvim-metals"] = function()
      dap.repl.open()
    end
```

**Reply by jtonic** [reply]
Is there a way to read user input from the console (dap-repl)?

Steps:
1. run the application using `run code lens`,  
2. open `dap-repl` window, and focus it,
3. go in INSERT mode to input data

Expected:
- Be able to input data and press <Enter>, and scala app read the inserted data and continue.

Actual:
- When enter in the input mode the `dap>` prompt appears to ask for debug information like variables, etc.
Application is blocked because waits for data to be inserted in the console. :(

I can run the application with `sbt` or `bsp` but I would like to have this simpler with the `nvim-metals` and `nvim-dap` support, especially when for mini CLI apps developed with `scala-cli`, no build tool whatsoever?  Moreover I would like to do the same but in debug mode.

Thanks in advance for any useful guidance.

**Reply by ckipp01** [reply]
When you set up your DAP configuration you can create one like this:

```lua
    dap.configurations.scala = {
      {
        type = "scala",
        request = "launch",
        name = "Run or test with input",
        metals = {
          runType = "runOrTestFile",
          args = function()
            local args_string = vim.fn.input("Arguments: ")
            return vim.split(args_string, " +")
          end,
        },
      }
    }
```

Then when you trigger this one it will prompt you for input before it actually launches.

**Reply by jtonic** [reply]
Thank you very much @ckipp01,

I think I was wrongfully expressed my case. Sorry!
What I am looking to achieve with nvim, nvim-metals, dap is something like in vs code / metals, as in image below, meaning
being able to input data in the terminal/console while the application is running, whenever the run. flow hits an input APIs (such as readLine).
When the dap REPL is displaying `What's your name?")` in the example, I want to be able to insert the information in the console, and after hitting `Enter` the application continues after the readLine() statement.
But when I tried to enter the insert mode, the dap repl prompted me with `dap>` shell.

I would like to be able to input information while running dap via run/debug commands or code lense.
Moreover I'd like to be able to input information (via readLine or alternatives) in whatever term/console (tmux), not necessary in the dap repl.

Thanks  

<img width="909" alt="image" src="https://github.com/user-attachments/assets/7e4c6f9a-b703-4de7-818d-5adc9241f63f">

**Reply by ckipp01** [reply]
Ahh ok, so you're a bit more limited than in VS Code here @jtonic but you can provide an initial prompt like this:

```
    dap.configurations.scala = {
      {
        type = "scala",
        request = "launch",
        name = "Run or test with input",
        metals = {
          runType = "runOrTestFile",
          args = function()
            local args_string = vim.fn.input("Arguments: ")
            return vim.split(args_string, " +")
          end,
        },
      }
}
```
This will at least allow you a prompt before running, but getting the readline thing like you mentioned multiple times isn't currently supported in nvim-dap I believe.

**Reply by doctorstal** [reply]
Hey @ckipp01, nice work here!
Is there a way to run and debug sbt project as in `sbt "project service" run`?
I can only see a possibility to launch an App

**Reply by ckipp01** [reply]
I think if that `service` has a main method, then yea it should be the same process.

**Comment by ckipp01** [comment]
# Creating a new Scala project

Under the hood Metals calls g8 to do this. You can use this with the New Scala Project command (`:MetalsNewScalaProject`).

![2021-12-08 19 58 24](https://user-images.githubusercontent.com/13974112/145267811-18bd2291-128e-437f-9b34-22cd27bd2a05.gif)

**Reply by sagoez** [reply]
Hello,

For some reason I can only make this work on a already existing Scala project. Is there a way to do it in a blank new git project?

**Reply by ckipp01** [reply]
Hey! So unfortunately nvim-metals needs to "attach" before the commands become available, meaning you already need to be in a Scala project to create a new one. It's preferable to just use `g8` directly if you don't have any project yet, which is what Metals uses under the hood.

**Reply by sagoez** [reply]
Got it, that makes sense. Thanks a lot!

**Comment by ckipp01** [comment]
### Creating a new Java File

Just like the feature to create a new Scala file, you can also trigger the "New Java File" command (`:MetalsNewJavaFile`) to get the option to create a new Java Class, Interface, or Enum.

![2021-12-29 13 32 35](https://user-images.githubusercontent.com/13974112/147662709-8facc608-9e24-4b1d-911b-48d3041762d2.gif)

**Comment by ckipp01** [comment]
### Show Build Target Info

Starting with 0.11.2 (or the latest snapshot as of posting this) you can view all sorts of info about a chosen build target by using the `:MetalsShowBuildTargetInfo` command. This will then give you a picker to choose which target you'd like to see, and then show you all the info.

![2022-02-21 20 12 56](https://user-images.githubusercontent.com/13974112/155016823-9bb39046-dbd7-49eb-858a-13bbce04cf8d.gif)

**Comment by ckipp01** [comment]
### Toggling the Metals Logs

This feature is used to show you the Metals-specific logs for your workspace.

Behind the scenes all this command does is issue a `tail -f .metals/metal.log` inside of a Neovim term.

![2022-12-29 07 19 02](https://user-images.githubusercontent.com/13974112/209911883-429e791f-2b31-40ff-951a-8425959b036a.gif)

**Comment by ckipp01** [comment]
### Run Scalafix Rules

#### NOTE: This feature currently requires usage of the latest Metals Snapshot `serverVersion = latest.snapshot`.

Running the `:MetalsRunScalafix` command will run the rules that are defined in your `.scalafix.conf` file. Note that external rules (apart from OrganizeImports) currently also require you to defined them in the `scalafixRulesDependencies` table in your Metals config.

![2022-06-11 14 50 16](https://user-images.githubusercontent.com/13974112/173188777-45bb99e9-9823-4c6c-b9a9-7b6b8d867bfe.gif)
